odoo/addons/base/models/ir_cron.py

```diff
@@ -290,91 +300,182 @@ def _acquire_one_job(cls, cr, job_ids):
+        (i.e., the server action notifies no remaining records to process)
+        or it fails without any record processed (i.e., the server action
+        crashes and notifies no record done.)
+        Each call to the server action is executed in its own transaction.
+        The server action is called up to ``MAX_BATCH_PER_CRON_CALL`` (10)
+        times in this loop.
+
+        :return: a string with 3 possible values:
+         - 'FULLY_DONE': There is no remaining record to process or no record has been processed
+         - 'PARTIALLY_DONE': Record have been processed and there is remaining record to process,
+                            even in case of exception
+         - 'FAILED': There was an exception during the execution and either no record has been processed or there is no
+                    remaining record to process.
+        """
+        timed_out_counter = job['timed_out_counter']
 
         with cls.pool.cursor() as job_cr:
-            lastcall = fields.Datetime.to_datetime(job['lastcall'])
-            interval = _intervalTypes[job['interval_type']](job['interval_number'])
-            env = api.Environment(job_cr, job['user_id'], {'lastcall': lastcall})
-            ir_cron = env[cls._name]
-
-            # Use the user's timezone to compare and compute datetimes,
-            # otherwise unexpected results may appear. For instance, adding
-            # 1 month in UTC to July 1st at midnight in GMT+2 gives July 30
-            # instead of August 1st!
-            now = fields.Datetime.context_timestamp(ir_cron, datetime.utcnow())
-            past_nextcall = fields.Datetime.context_timestamp(
-                ir_cron, fields.Datetime.to_datetime(job['nextcall']))
-
-            # Compute how many call were missed
-            missed_call = past_nextcall
-            missed_call_count = 0
-            while missed_call <= now:
-                missed_call += interval
-                missed_call_count += 1
-            future_nextcall = missed_call
-
-            # Compute how many time we should run the cron
-            effective_call_count = (
-                     1 if not missed_call_count                    # run at least once
-                else 1 if not job['doall']                         # run once for all
-                else missed_call_count if job['numbercall'] == -1  # run them all
-                else min(missed_call_count, job['numbercall'])     # run maximum numbercall times
-            )
-            call_count_left = max(job['numbercall'] - effective_call_count, -1)
+            env = api.Environment(job_cr, job['user_id'], {'lastcall': job['lastcall'], 'cron_id': job['id']})
+            cron = env[cls._name].browse(job['id'])
+            status = None
+            for i in range(MAX_BATCH_PER_CRON_JOB):
+                cron = cron._add_progress(timed_out_counter=timed_out_counter)
+                progress = cron._get_progress()
```

Comment:
Could be nice that `_add_progress` return the progress as second element so we could write
```suggestion
                cron, progress = cron._add_progress(timed_out_counter=timed_out_counter)
```

Reply:
That was the initial proposal, and I personally find this "dumb" version much clearer.